Src/GoogleApis.Auth.DotNet4/OAuth2/JwtServiceAccountCredential.cs

```diff
@@ -0,0 +1,121 @@
+    /// </para>
+    /// </summary>
+    public class JwtServiceAccountCredential : ServiceAccountCredential
+    {
+        /// <summary>Default lifetime of 1 hour</summary>
+        private const int DefaultExpiresInSeconds = 3600;
+
+        /// <summary>Constructs a new JWT service account credential using the given initializer.</summary>
+        public JwtServiceAccountCredential(ServiceAccountCredential.Initializer initializer)
+            : base(initializer)
+        {
+        }
+
+        /// <summary>
+        /// Gets an access token to authorize a request.
+        /// If <paramref name="authUri"/> is set and this credential has no scopes associated
+        /// with it, a JWT access token for given <paramref name="authUri"/> is returned.
+        /// Otherwise, the access token will be obtained in exactly the same way as in
+        /// <see cref="ServiceAccountCredential"/>
+        /// </summary>
+        /// <param name="authUri">The URI the returned token will grant access to.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The access token.</returns>
+        public override async Task<string> GetAccessTokenForRequestAsync(string authUri = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (!HasScopes && authUri != null)
+            {
+                // TODO(jtattermusch): support caching of JWT access tokens per authUri, currently a new 
+                // JWT access token is created each time, which can hurt performance.
+                return CreateJwtAccessToken(authUri).AccessToken;
+            }
+            return await base.GetAccessTokenForRequestAsync(authUri, cancellationToken);
+        }
+
+        /// <summary>
+        /// Creates a JWT access token than can be used in request headers instead of an OAuth2 token.
+        /// This is achieved by signing a special JWT using this service account's private key.
+        /// <param name="authUri">The URI for which the access token will be valid.</param>
+        /// </summary>
+        private JwtAccessToken CreateJwtAccessToken(string authUri)
+        {
+            var issuedDateTime = Clock.UtcNow;
+            var issued = (int)(issuedDateTime - UnixEpoch).TotalSeconds;
+            var payload = new JsonWebSignature.Payload()
+            {
+                Issuer = Id,
+                Subject = Id,
+                Audience = authUri,
+                IssuedAtTimeSeconds = issued,
+                ExpirationTimeSeconds = issued + DefaultExpiresInSeconds,
+            };
+
+            return new JwtAccessToken
+            {
+                Issued = issuedDateTime,
+                ExpiresInSeconds = DefaultExpiresInSeconds,
+                AccessToken = CreateAssertionFromPayload(payload)
+            };
+        }
+
+        /// <summary>
+        /// JWT access token
+        /// </summary>
+        internal class JwtAccessToken
```

Comment:
Why can't you use https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis.Auth/OAuth2/Responses/TokenResponse.cs?